### ISSUE TITLE
TS function declaration syntax does not cause errors in JS files

### ISSUE THREAD

**Issue by Josh-Cena** [microsoft/TypeScript#54399]
# Bug Report

### 🔎 Search Terms

Can only be used in TypeScript files; TSDeclareFunction

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=7&ssc=4&pln=1&pc=1&filetype=js#code/PQKhCgAIUgVALAlgZ0gKQMoBED2BjSFSAQwBNSBTUyZHQgF0gHccAnAa1QoDcKA7ZonrxIeeBTzs0yAHRQYAAVYV6AV1Z9UAb1V92fHEz4BfecHAAzXXnqIcAiwAoAlAG5wlgITugA)

### 💻 Code

```ts
/**
 * This JSDoc is added so it works even with checkJs.
 * @returns {unknown}
 */
function f();

f!;
```

### 🙁 Actual behavior

`function f();` is accepted but it's invalid JS.

### 🙂 Expected behavior

Gives the same error as the non-null assertion below that it can only be used in TypeScript files.
